src/data/int/log.lean

```diff
@@ -0,0 +1,245 @@
+--    (0,                      1,                     1)
+```
+which means it can be used for computing digit expansions
+```lean
+import data.fin.vec_notation
+
+def digits (b : ℕ) (q : ℚ) (n : ℕ) : ℕ :=
+⌊q*b^(↑n - int.log b q)⌋₊ % b
+
+#eval digits 10 (1/7) ∘ (coe : fin 8 → ℕ)
+-- ![1, 4, 2, 8, 5, 7, 1, 4]
+```
+
+## Main results
+
+* For `int.log`:
+  * `int.zpow_log_le_self`, `int.lt_zpow_succ_log_self`: the bounds formed by `int.log`,
+    `(b : R) ^ log b r ≤ r < (b : R) ^ (log b r + 1)`.
+  * `int.zpow_le_iff_le_log`: the galois almost-connection
+* For `int.clog`:
+  * `int.zpow_pred_clog_lt_self`, `int.self_le_zpow_clog`: the bounds formed by `int.log`,
+    `(b : R) ^ (clog b r - 1) < r ≤ (b : R) ^ clog b r`.
+  * `int.le_zpow_iff_clog_le`: the galois almost-connection
+* `int.neg_log_inv`: the link between the two definitions
+
+-/
+variables {R : Type*} [linear_ordered_field R] [floor_ring R]
+
+namespace int
+
+/-- The greatest power of `b` such that `b ^ log b r ≤ r`. -/
+def log (b : ℕ) (r : R) : ℤ :=
+if 1 ≤ r then
+  nat.log b ⌊r⌋₊
+else
+  -nat.clog b ⌈r⁻¹⌉₊
+
+lemma log_of_one_le_right (b : ℕ) {r : R} (hr : 1 ≤ r) : log b r = nat.log b ⌊r⌋₊ :=
+if_pos hr
+
+lemma log_of_right_le_one (b : ℕ) {r : R} (hr : r ≤ 1) : log b r = -nat.clog b ⌈r⁻¹⌉₊ :=
+begin
+  obtain rfl | hr := hr.eq_or_lt,
+  { rw [log, if_pos hr, inv_one, nat.ceil_one, nat.floor_one, nat.log_one_right, nat.clog_one_right,
+        int.coe_nat_zero, neg_zero], },
+  { exact if_neg hr.not_le }
+end
+
+@[simp, norm_cast] lemma log_nat_cast (b : ℕ) (n : ℕ) : log b (n : R) = nat.log b n :=
+begin
+  cases n,
+  { simp [log_of_right_le_one _ _, nat.log_zero_right] },
+  { have : 1 ≤ (n.succ : R) := by simp,
+    simp [log_of_one_le_right _ this, ←nat.cast_succ] }
+end
+
+lemma log_of_left_le_one {b : ℕ} (hb : b ≤ 1) (n : R) : log b n = 0 :=
+begin
+  cases le_total 1 n,
+  { rw [log_of_one_le_right _ h, nat.log_of_left_le_one hb, int.coe_nat_zero] },
+  { rw [log_of_right_le_one _ h, nat.clog_of_left_le_one hb, int.coe_nat_zero, neg_zero] },
+end
+
+lemma log_of_right_le_zero {n : R} (hn : n ≤ 0) (b : ℕ) : log b n = 0 :=
+by rw [log_of_right_le_one _ (hn.trans zero_le_one),
+    nat.clog_of_right_le_one ((nat.ceil_eq_zero.mpr $ inv_nonpos.2 hn).trans_le zero_le_one),
+    int.coe_nat_zero, neg_zero]
+
+lemma zpow_log_le_self {b : ℕ} {r : R} (hn : 1 < b) (hr : 0 < r) :
+  (b : R) ^ log b r ≤ r :=
+begin
+  cases le_total 1 r with hr1 hr1,
+  { rw log_of_one_le_right _ hr1,
+    refine le_trans _ (nat.floor_le hr.le),
+    rw [zpow_coe_nat, ←nat.cast_pow, nat.cast_le],
+    exact nat.pow_log_le_self hn (nat.floor_pos.mpr hr1) },
+  { rw [log_of_right_le_one _ hr1, zpow_neg, zpow_coe_nat, ← nat.cast_pow],
+    apply inv_le_of_inv_le hr,
+    refine (nat.le_ceil _).trans (nat.cast_le.2 _),
+    exact nat.le_pow_clog hn _ },
+end
+
+lemma lt_zpow_succ_log_self (b : ℕ) (r : R) (hn : 1 < b) (hr : 0 < r) :
```

Comment:
```suggestion
lemma lt_zpow_succ_log_self {b : ℕ} {r : R} (hn : 1 < b) (hr : 0 < r) :
```
Would also be much cleaner if you declared them once and for all with `variables {b : ℕ} {r : R}` at the top of the file.